Spigot-Server-Patches/0747-Add-feature-access-to-custom-ChunkGenerators.patch

```diff
@@ -0,0 +1,167 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: dfsek <dfsek@protonmail.com>
+Date: Tue, 25 May 2021 17:09:34 -0700
+Subject: [PATCH] Add feature access to custom ChunkGenerators
+
+
+diff --git a/src/main/java/io/papermc/paper/world/CraftProtoWorld.java b/src/main/java/io/papermc/paper/world/CraftProtoWorld.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..e037126f1fd2af9bd196c57e9af5856dc6657318
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/world/CraftProtoWorld.java
+@@ -0,0 +1,97 @@
++package io.papermc.paper.world;
++
++import com.google.common.base.Preconditions;
++import net.minecraft.core.BlockPosition;
++import net.minecraft.server.level.RegionLimitedWorldAccess;
++import net.minecraft.world.entity.EntityInsentient;
++import net.minecraft.world.entity.EnumMobSpawn;
++import net.minecraft.world.level.block.BlockFluids;
++import org.bukkit.World;
++import org.bukkit.block.data.BlockData;
++import org.bukkit.craftbukkit.block.data.CraftBlockData;
++import org.bukkit.entity.Entity;
++import org.bukkit.event.entity.CreatureSpawnEvent;
++import org.bukkit.util.Consumer;
++import org.bukkit.util.Vector;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.lang.ref.WeakReference;
++import java.util.Objects;
++
++public class CraftProtoWorld implements ProtoWorld {
++    /*
++     * Keep the world in a weak reference to prevent people from doing dumb
++     * things that keep RLWA's around after we're done with them.
++     */
++    private final WeakReference<RegionLimitedWorldAccess> world;
++
++    public CraftProtoWorld(RegionLimitedWorldAccess world) {
++        this.world = new WeakReference<>(world);
++    }
++
++    /**
++     * Clears the {@link WeakReference} to the RLWA.
++     */
++    public void clearReference() {
++        world.clear();
++    }
++
++
++    @Override
++    public void setBlockData(int x, int y, int z, @NotNull BlockData data, boolean updateLiquids) {
++        // RLWA does not support physics updates, so we only need to worry about scheduling fluid updates.
++        BlockPosition position = new BlockPosition(x, y, z);
++        getDelegate().setTypeAndData(position, ((CraftBlockData) data).getState(), updateLiquids ? 3 : 2 | 16 | 1024);
++        if(updateLiquids && ((CraftBlockData) data).getState().getBlock() instanceof BlockFluids) {
```

Comment:
there are many reasons people will want to queue updates, it should not be restricted to just liquids (and this is done incorrectly anyways, doesn't account for waterloggable). Should add different dedicated functions that let people queue fluid or block updates for positions.

You will be able to use the getFluid method from RWLA to get the fluid state for a position.